source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2020 André Perez Maselco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_inline_function.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const spvtools::fuzz::protobufs::TransformationInlineFunction& message)
+    : message_(message) {}
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const std::map<uint32_t, uint32_t>& result_id_map,
+    uint32_t function_call_id) {
+  *message_.mutable_result_id_map() =
+      google::protobuf::Map<google::protobuf::uint32, google::protobuf::uint32>(
+          result_id_map.begin(), result_id_map.end());
+  message_.set_function_call_id(function_call_id);
+}
+
+bool TransformationInlineFunction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // The values in the |message_.result_id_map| must be all fresh and all
+  // distinct.
+  std::set<uint32_t> ids_used_by_this_transformation;
+  for (auto& pair : message_.result_id_map()) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+            pair.second, ir_context, &ids_used_by_this_transformation)) {
+      return false;
+    }
+  }
+
+  // |function_call_instruction| must be defined, must be an OpFunctionCall
+  // instruction, must be the penultimate instruction in its block and its block
+  // termination instruction must be an OpBranch.
```

Comment:
I see that we don't and that's good. One particular issue with this case is that once we inline the function, we might end up with a situation where a back-edge of the loop does not point to the loop's header.

It might be good to make an explicit comment about this and add a test for it.

Reply:
@Vasniktel I don't quite understand. Can you give an example?

Reply:
I've been thinking about this situation:
```
%1 = OpFunction %int None %type
%2 = OpLabel
OpBranch %3
%3 = OpLabel
OpReturnValue %int_constant
OpFunctionEnd
...
%4 = OpLabel
%5 = OpFunctionCall %int %1
OpLoopMerge ....
OpBranch %6
...
```
If we inline `%1`, we might get:
```
%4 = OpLabel
OpBranch %7
%7 = OpLabel
%8 = OpCopyObject %int %inst_constant
OpLoopMerge ...
OpBranch %6
...
```
In this situation, there is a back-edge block somewhere branching to `%4` which is no longer a loop header.

Reply:
Oh I see. You mean if we did not require the call to be the penultimate instruction then we could get the above situation. Yes, it is good that we avoid this. No action is needed for this comment chain, right?

Reply:
Yeah, it would be good to comment and test it, though.

Reply:
Ah right, yes you said that. Sounds good. E.g. a comment like "This avoids the case where..." and also a test that tries to inline a function call at the end of a loop header.

Reply:
Done.